### PR TITLE
Expose ExpressJS functionality of `next('router')`

### PR DESCRIPTION
**Pull request by hermanbanken**
Quoting [documentation](https://expressjs.com/en/guide/using-middleware.html#middleware.router):

<blockquote>
To skip the rest of the router’s middleware functions, call next('router') to pass control back out of the router instance.

This example shows a middleware sub-stack that handles GET requests to the /user/:id path.
<pre language="javascript">
var app = express()
var router = express.Router()

// predicate the router with a check and bail out when needed
router.use(function (req, res, next) {
  if (!req.headers['x-auth']) return next('router')
  next()
})

router.get('/user/:id', function (req, res) {
  res.send('hello, user!')
})

// use the router and 401 anything falling through
app.use('/admin', router, function (req, res) {
  res.sendStatus(401)
})
</pre></blockquote>

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] https://expressjs.com/en/guide/using-middleware.html#middleware.router
- [x] This has been in ExpressJS [for a long time, see commit](https://github.com/expressjs/express/commit/9722202df964bfbfc0f579e4baeb5a4e1b43b344), at least 5.0.0-alpha.7 5.0.0-alpha.6 5.0.0-alpha.5 5.0.0-alpha.4 4.17.1 4.17.0 4.16.4 4.16.3 4.16.2 4.16.1 4.16.0 4.15.5 4.15.4 4.15.3 4.15.2 4.15.1 4.15.0
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

File: types/express-serve-static-core/index.d.ts
Comment (peterblazejewicz) [inline types/express-serve-static-core/index.d.ts:38]: I'm pretty sure the checks are now startin from 2.8, so that lint is no longer required here to siable tslint `callbacl-types` rule

**Review by weswigham — CHANGES_REQUESTED** [review]
Requesting @peterblazejewicz 's changes

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@hermanbanken thx for the contribution!

**Review by weswigham — APPROVED** [review]

**Comment by hermanbanken** [comment]
Ready to merge

**Comment by kawanet** [comment]
@hermanbanken @weswigham This is a revert request.
After the PR, now `tsc` raises many errors on my app which is not changed.
Before the PR, `NextFunction` did not require types both for arguments and return value.
After the PR, it requires types. I need to make ugly changes like below:

```diff
-    handler(req, res, e => (e ? reject(e) : resolve()));
+    handler(req, res, ((e?: any):void => (e ? reject(e) : resolve()));
```

I'm sorry but I'm rarely sure why the change is needed as the `any` type could include the string of `"router"` as well.

```diff
export interface NextFunction {
     (err?: any): void;
+    (deferToNext: "router"): void; // needless?
}
```

`tsc` raises errors like below:

- TS2345: Argument of type '(err?: Error) => void' is not assignable to parameter of type 'NextFunction'. Types of parameters 'err' and 'deferToNext' are incompatible. Type '"router"' is not assignable to type 'Error'.
- TS7006: Parameter 'err' implicitly has an 'any' type.
- TS7011: Function expression, which lacks return-type annotation, implicitly has an 'any' return type.

**Comment by weswigham** [comment]
I defer to @peterblazejewicz and the other package maintainers~

**Comment by peterblazejewicz** [comment]
@kawanet
I see the problem, sorry for not spotting this promtly. I'll add a test case for regression.
@hermanbanken
Can we retract this change (or refine)?  I see the obvious gains (it helps type), but I think we missed existing versatility being broken (as not test covered that case).

**Comment by hermanbanken** [comment]
Yes, let’s adapt it to be compatible. I’m not 100% sure how to combine the two syntaxes while keeping the relevant type information.

Any is too vague: the behavior is VERY different in a router. In fact, the reporter of this new bug might have a problem if someone throws “router”, which is possible as you can throw strings.

I’m off today, so I’m not able to make a change as I don’t have access to my computer.

If someone can add the test first and check how we can make the signature compatible, that would be great. Otherwise I will take it up on Tuesday.

**Pull request merged**
